app/domain/authentication/resource_restrictions/validate_resource_restrictions.rb

```diff
@@ -0,0 +1,78 @@
+require 'command_class'
+
+module Authentication
+  module ResourceRestrictions
+
+    # This command class validates the resource restrictions retrieved from the
+    # @extract_resource_restrictions command class dependency.
+    #
+    # Its inputs are:
+    # @authenticator_name - the name of the authenticator. e.g. authn-azure.
+    # @service_id - the specific service ID requested to authenticate. So if
+    #   the authenticator is `authn-azure/prod`, then `prod` is the service ID.
+    # @host_name - the name of the requesting Conjur host as defined in the DB.
+    # @account - the Conjur account where the relevant entities are defined.
+    # @constraints - an object that responds to validate(resource_restrictions)
+    #   where resource_restrictions is an array of restrictions found for the
+    #   requesting host. Usually will be `Constraints::MultipleConstraint`.
+    # @authentication_request - an object that each authenticator needs to
+    #   define. It should respond to retrieve_attribute(attribute_name) and
+    #   return a value to use for comparison with the corresponsing value
+    #   defined in the DB for the requesting host.
+    ValidateResourceRestrictions = CommandClass.new(
+      dependencies: {
+        extract_resource_restrictions: ExtractResourceRestrictions.new,
+        logger:                        Rails.logger
+      },
+      inputs:   %i(authenticator_name service_id host_name account constraints authentication_request)
+    ) do
+
+      def call
+        @logger.debug(LogMessages::Authentication::ValidatingResourceRestrictions.new(@host_name))
+
+        extract_resource_restrictions
+        validate_resource_restrictions_configuration
+        validate_request_matches_resource_restrictions
+
+        @logger.debug(LogMessages::Authentication::ValidatedResourceRestrictions.new)
+      end
+
+      private
+
+      def extract_resource_restrictions
+        resource_restrictions
+      end
+
+      def resource_restrictions
+        @resource_restrictions ||= @extract_resource_restrictions.call(
+          authenticator_name: @authenticator_name,
+          service_id:         @service_id,
+          host_name:          @host_name,
+          account:            @account
+        )
+      end
+
+      def validate_resource_restrictions_configuration
+        @logger.debug(LogMessages::Authentication::ValidatingResourceRestrictionsConfiguration.new)
+
+        @constraints.validate(
+          resource_restrictions: resource_restrictions.names
+        )
+
+        @logger.debug(LogMessages::Authentication::ValidatedResourceRestrictionsConfiguration.new)
+      end
+
+      def validate_request_matches_resource_restrictions
+        @logger.debug(LogMessages::Authentication::ValidatingResourceRestrictionsValues.new)
+
+        resource_restrictions.each do |restriction_name, restriction_value|
+          if @authentication_request.retrieve_attribute(restriction_name) != restriction_value
```

Comment:
this line is a bit clear and took me some time to understand the `restriction_value` is basically the value of the constants from the resulting restriction_name. Therefore, I think maybe `(annotation_restriction_name) != constant_restriction_name` is more clear because otherwise in my mind at least name and value are key -> value pairs

Reply:
> in my mind at least name and value are key -> value pairs

It *is* a key-value pair.
I put a lot of effort to differentiate between 4 common confusing terms:
1. resource restrictions
2. constraints
3. annotations
4. request attributes

The way I separated them is:
- Annotations: as defined in the policy (not processed)
- Resource Restrictions: key-value mapping of restrictions that the request must adhere to to authenticate.
  It is derived from the Annotations by selecting only the ones begin with `authenticator-name/`, without the prefix.
  Its value is the annotation value.
- Constraints: list of rules that the Resource Restrictions needs to adhere to to be considered valid restrictions.
  This validation step can (and should) be done in policy load.
- Request attributes: should not be confused with Resource Restrictions, these are attributes of the request that are needed to validate the request.
  They are retrieved differently for each authenticator (JWT token/K8s API or else).
  The request is validated by making sure they adhere to the Resource Restrictions.
  That is, for each resource restriction, the restriction value equals the corresponding request attribute value.

Reply:
@abrahamko does this data need to be written somewhere? will it help future devs?

Reply:
@orenbm @sigalsax Is there a doc or something I can put this info in?

Reply:
So if I am understanding correctly, resource restrictions is intended to be a hash map of acceptable_resource_restriction -> user_given_resource_restriction ?

Reply:
> Is there a doc or something I can put this info in?

Well to start, it should be in the code if not already and I think it should be given it's own category or own link here: https://github.com/cyberark/conjur/tree/master/design. It could go under it's own title "Things to be aware when creating a design" or "common confusions" or "key terms"